File: Design Pattern/11/abstract.java
public abstract class Beverage{
  final void init(){
    boilWater();
    brew();
    pourIncup();
    addCondiments();
  }
  void boilWater(){
    System.out.println("Boil water");
  }
  abstract void brew();
  abstract void addCondiments();
  abstract void pourIncup();
}

public class Coffee extends Beverage{
//重载函数
  void brew(){
    System.out.println("brewing Coffee");
  }
//重载函数
  void pourIncup(){
    System.out.println("pouring Coffee");
  }
//重载函数
  void addCondiments(){
    System.out.println("Add suger and milk");
  }
}

public class Tea extends Beverage{
//重载函数
  void brew(){
    System.out.println("brewing tea");
  }
//重载函数
  void pourIncup(){
    System.out.println("pouring tea");
  }
//重载函数
  void addCondiments(){
    System.out.println("Add lemon");
  }
}


public lass Test{
  public static void main(){
    Beverage coffee=new Coffee();
    coffee.init();
    Beverage tea=new Tea();
    tea.init();
  }
}

File: Design Pattern/11/abstract.js
var Beverage=function (param) {
  var boilWater=function () {
    console.log("BoilWater...");
  }
  var brew=param.brew||function () {
    console.log("Brewing ....");
  }
  var pourIncup=param.pourIncup||function () {
    console.log("pourIncup....");
  }
  var addCondiments=param.addCondiments||function () {
    console.log("addCondiments...");
  }

  var f=function(){};
  f.prototype.init=function () {
    boilWater();brew();
    pourIncup();addCondiments();
  }
  return f;
}

var Coffer=Beverage({
  brew:function () {
    console.log("brewing coffer");
  },
  addCondiments:function () {
    console.log("addCondiments coffer");
  }
});


var tea=Beverage({
  brew:function () {
    console.log("brewing tea");
  },
  addCondiments:function () {
    console.log("addCondiments tea");
  }
});

var c=new Coffer();
c.init()
